crypto/ec/ec_pmeth.c

```diff
@@ -17,7 +17,7 @@
 #include "internal/evp_int.h"
 
 #if !defined(OPENSSL_NO_SM2)
-# include <openssl/sm2.h>
+# include "internal/sm2.h"
```

Comment:
I haven't looked at original SM2 submission, but I kind of struggle with and leaning toward objecting the way SM2 is "wired" into ec_pmeth.c. And rather "hacked" into than "wired", in vocabulary meaning of the term "hack". It should either be implemented as own pmeth, or have own kmeth that would override default, or something else. Incidentally, recent comments #4793 suggest that it would be appropriate to do something about it.

Reply:
The initial implementation was using its own pmeth but it was removed along the way because wiring into existing `ec_pmeth.c` seemed simpler and without harm. But indeed the comments in 4793 suggest this was a mistake.

Reply:
It's not a big deal and can be fixed now, as a bug-fix, over the next couple of weeks?

Reply:
OK. I will create a PR moving SM2 to its on PMETH in next week or so.

Reply:
Even though #4793 suggests that SM2 is specified over single curve, it doesn't actually mean that implementation may not allow parameterization. In sense that if tomorrow they approve another curve, we should not find ourselves in situation that we would have to maintain it as [another] special case. And here I sense ambiguity of following character. As it stands now the SM2 is engaged based on curve's[!] nid. But it's just another curve and it doesn't actually define SM2 scheme, in sense that the scheme is oblivious of the actual curve. So it would be more than appropriate to engage it by means other than curve's nid.

This also calls for following question. Customarily signature would be placed into an ASN.1 structure, that would describe algorithm and parameters. Are there OIDs that identify SM2 *scheme* [and its parameters]? Should they be recognized [and threated accordingly]?

Reply:
@dot-asm I agree with the suggestion, and certainly the SM2 algorithms themselves are not locked to a single curve. We should allow that flexibility indeed.

However, for clarity purposes, allow me to reiterate that there is currently _only one SM2 curve_. This specific curve is provided in GB/T 32918.5, as part of the GB/T 32918 standard that defines SM2 itself. It is currently the _only compliant curve_ to the SM2 standard.

This is @InfoHunter 's page on OIDs: https://github.com/InfoHunter/misc/blob/master/GBT33560-2017-A1.md . The SM2 signature scheme OIDs have already been implemented in previous PRs.

Reply:
I think I saw @dot-asm suggest that ec_curve.c could be changed to include specific EC_KEY_METHODs with select curves, and thereby hook the SM2 curves with the SM2 algos by default.  I know I was thinking something like that as well early on when looking at #4793, but couldn't quite get it together without making things hugely incompatible with 1.1.0...  but maybe there's a possibility to do this?

Reply:
I see only 301 in objects.txt and it's apparently used to identify the curve itself. (If assertion is that it identifies whole scheme, then one can wonder if it's appropriate to use it to describe the curve.) Question rather is if there should even be 501(?) that should engage SM2 when parsing corresponding ASN.1 object.

Reply:
301 represents the SM2 scheme. There is no official OID that represents the SM2 curve. I believe we went with 301 for compatibility reasons since some other implementations used 301.

Reply:
This doesn't make ambiguity less ambiguous...